### PR TITLE
feat: experimental "dev mode" with much better container syncing

### PR DESCRIPTION
**Pull request by edvald**
This introduces a new notion of _dev mode_ for services, along with a complete replacement for the current _hot reloading_ mechanism, based on the excellent [Mutagen](https://github.com/mutagen-io/mutagen) project by @xenoscopic, which is *vastly* superior to the prior rsync-based implementation in performance, and allows for two-way syncing as well.

I'm very excited by how well this works in my testing, but because this is such a major change and also introduces new semantics and configuration layouts, this will be flagged as experimental for a bit. The older hot reloading mechanism will still be around, but will eventually be deprecated and completely removed.

To use the feature, services need to be configured with the `devMode` field, which is similar to the existing `hotReload` field in shape, but is configured per-service instead of just at the `container` module level. 

Couple of examples from the vote and vote-helm example projects:

```yaml
kind: Module
description: The voting UI
name: vote
type: container
services:
  - name: vote
    args: [npm, run, serve]
    devMode:
      sync:
        - target: /app/src
          source: src
          mode: two-way
...
```

```yaml
kind: Module
description: Helm chart for the voting UI
type: helm
name: vote
dependencies:
  - api

serviceResource:
  kind: Deployment
  containerModule: vote-image
devMode:
  sync:
    - target: /app/src
      source: src
      mode: two-way
```

Notice the new `mode` field, which allows enabling two-way sync.

To run services in dev mode, once configured, you can simply run `garden dev` as usual, `garden dev <service name>`, or `garden deploy --dev=<service-name>`.

In order to make the Mutagen integration work, we're temporarily running a fork of it, which includes a new transport protocol. Hopefully it'll make its way upstream soon.

TODO
- [x] Integration tests
- [x] Documentation
- [x] Test on Windows+Linux
- [ ] Support postSyncCommand in dev mode (I left it off the devMode configuration schema for now)

### CODE REVIEW

**Comment by xenoscopic** [comment]
Happy to help with what I can here.  I've reviewed the Garden fork of Mutagen and that implementation seems sound.  Feel free to ping me if there are any specific questions, and definitely watch out for the v0.12 release of Mutagen in the near future, as it would be a better version to serve as a baseline for rebasing.

**Review by eysi09 — COMMENTED** [review]
Had a few comments, otherwise looks 🔥! Great job!

I also gave it a spin on my end with a handful of services in two-way dev mode at the same time, and everything worked as expected.

File: core/src/analytics/analytics.ts [hunk 1]
```diff
@@ -134,9 +133,6 @@ export class AnalyticsHandler {
   private projectMetadata: ProjectMetadata
 
   private constructor(garden: Garden, log: LogEntry) {
-    if (!garden.sessionId) {
```
[the remark this answers is not in the document]
Comment (thsig) [reply core/src/analytics/analytics.ts:143]: Yeah, that's right.

File: core/test/integ/src/plugins/kubernetes/container/deployment.ts [hunk 2]
```diff
@@ -374,6 +469,69 @@ describe("kubernetes container deployment handlers", () => {
...
+          service,
+          runtimeContext: emptyRuntimeContext,
+          log,
+          devMode: true,
+          hotReload: false,
+        })
+
+        // First, we create a file locally and verify that it gets synced into the pod
+        await writeFile(join(module.path, "made_locally"), "foo")
+        await sleep(300)
```
[the remark this answers is not in the document]
Comment (thsig) [reply core/test/integ/src/plugins/kubernetes/container/deployment.ts:501]: True that. I think 300ms is very generous, so I'd be surprised if we run into problems here (we'll see, of course).

File: core/test/e2e/src/pre-release.ts [hunk 3]
```diff
@@ -98,8 +98,8 @@ describe("PreReleaseTests", () => {
   if (project === "demo-project") {
     describe("demo-project", () => {
       describe("top-level sanity checks", () => {
-        it("runs the dev command", async () => {
```
[the remark this answers is not in the document]
Comment (thsig) [reply core/test/e2e/src/pre-release.ts]: I've re-enabled the tests for the dev command.

**Review by eysi09 — APPROVED** [review]

**Comment by ArnaudRinquin** [comment]
👀

**Pull request merged**